### PR TITLE
[DF] Fix Snapshot of RVecs read from a TTree

### PR DESCRIPTION
**Pull request by eguiraud**
Before this commit, the logic in SnapshotHelper did not take into
account the case of RVecs that are being read from an input TTree
(that was not possible until v6.26). Because of how the logic was
structured, that meant that Snapshot was trying to write out those
RVecs as C-style arrays instead, obviously badly failing at it.

This patch tries to make the flow of this logic a bit more readable
and fixes the problem mentioned above by treating input branches
of RVec type the same way we already treat std::vectors.

This fixes https://github.com/root-project/root/issues/10225.

@pcanal the logic in this commit could be simplified if instead of "if it's not every other case, it must be a C-style array" I could do instead "if it's not a C-style array, then it must be one of the other cases". But I'm not sure how to check for "branch is a C-style array", see https://mattermost.web.cern.ch/root/pl/4n1qbh363tdfjc1hs3khe6y84o

### CODE REVIEW

**Comment by pcanal** [comment]
```
TBranchElement *be = dynamic_cast<TBranchElement*>(branch_ptr);
bool isarray = (be && be->GetInfo()->GetElement(be->GetID())->GetArrayDim() > 0);
```
and
```
if (!be) // && and !dynamic_cast<TBranchObject*>(branch_ptr)
{
    bool isarray = branch->GetListOfLeaves()->GetEntries() == 1 && (branch->GetListOfLeaves()->At(0)->GetLeafCount() != nullptr || branch->GetListOfLeaves()->At(0)->GetLenStatic() > 1);
}
```
and for TBranchObject, we probably need to look at `TLeafObject::GetTypeName` (or not support it)

**Comment by eguiraud** [comment]
Ok, that seems more complex than the current logic :sweat_smile:

File: tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx [hunk 1]
```diff
@@ -1343,21 +1344,30 @@ void SetBranchesHelper(TTree *inputTree, TTree &outputTree, const std::string &i
...
-   const bool isTClonesArray = inputBranch != nullptr && std::string(inputBranch->GetClassName()) == "TClonesArray";
-   const auto mustWriteRVec = !inputBranch || isTClonesArray ||
-                              ROOT::ESTLType::kSTLvector == TClassEdit::IsSTLCont(inputBranch->GetClassName());
+
+   // if no backing input branch, we must write out an RVec
+   bool mustWriteRVec = (inputBranch == nullptr);
+   // otherwise, if input branch is TClonesArray, must write out an RVec
+   if (!mustWriteRVec && std::string_view(inputBranch->GetClassName()) == "TClonesArray") {
+      mustWriteRVec = true;
+      Warning("Snapshot",
```
Comment (Axel-Naumann) [inline tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx:1353]: Do you have a test for this case?

**Review by Axel-Naumann — APPROVED** [review]
LGTM

**Review by pcanal — APPROVED** [review]
We should probably also have (add if not there already) a test of what happens if trying to read other STL collection into a RVec.

File: tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx (hunk 1 above)
Comment (eguiraud) [reply tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx:1353]: yes, it's here: https://github.com/root-project/root/blob/master/tree/dataframe/test/dataframe_snapshot.cxx#L782 (I'm not introducing this warning in this PR)

**Comment by eguiraud** [comment]
@pcanal I quickly tried with `std::array<int, 4>` but it looks like TTreeReaderArray has trouble with that type:

```cpp
#include <ROOT/RDataFrame.hxx>
#include <array>

int foo() {
  {
    std::array<int, 4> a{1, 2, 3, 4};
    TFile f("f.root", "recreate");
    TTree t("t", "t");
    t.Branch("a", &a);
    t.Fill();
    t.Write();
  }

  auto c = ROOT::RDataFrame("t", "f.root")
               .Filter([](std::array<int, 4> &a) { return a[0] == 1; }, {"a"})
               .Filter(
                   [](ROOT::RVecI &v) {
                     return v.size() == 4 && v[0] == 1 && v[3] == 4;
                   },
                   {"a"})
               .Count()
               .GetValue();
  return c;
}
```

yields

```
/tmp root -l -b -q foo.cpp+

Processing foo.cpp+...
Info in <TUnixSystem::ACLiC>: creating shared library /tmp/./foo_cpp.so
Error in <TTreeReaderValueBase::GetBranchDataType()>: Must use TTreeReaderArray to read branch a: it contains an array or a collection.
Error in <TTreeReaderValueBase::CreateProxy()>: The branch a contains data of type {UNDETERMINED TYPE}, which does not have a dictionary.
terminate called after throwing an instance of 'std::runtime_error'
  what():  An error was encountered while processing the data. TTreeReader status code is: 6
```

I'll put it on my to-do list to check.

As I mentioned above the safest way to fix this would be to switch from "if this is not a Define, or a std::vector, or an RVec, or a TClonesArray, then it must be a C-style array and it needs special care" to "if this is not a C-style array, then no special care needed", but even with your help I could not figure out how to code "is this a C-style array?".

**Comment by eguiraud** [comment]
Finally got around to check this and the situation is pretty much ok.

std::vector, std::array and std::deque branches can all be written out from a Snapshot correctly as those same types. Only std::vector can be written out as an RVec as well, with Snapshot doing the conversion on the fly.

There are a few warts in error reporting for unsupported cases but that's another issue.

**Pull request merged**